### PR TITLE
Support type coercion for timestamp and utf8

### PR DESCRIPTION
**Pull request by andre-cc-natzka**
# Which issue does this PR close?

Closes [#4311](https://github.com/apache/arrow-datafusion/issues/4311).

# Rationale for this change

Currently DataFusion supports type coercion for (`DataType::Time`, `DataType::Utf8`) and (`DataType::Date`, `DataType::Utf8`) pairs, but is still missing type coercion for (`DataType::Timestamp`, `DataType::Utf8`).

# What changes are included in this PR?

Two lines of code are added in the `temporal_coercion` function from `datafusion/expr/src/type_coercion/binary.rs` to account for type coercion for a (`DataType::Timestamp`, `DataType::Utf8`) pair. The output type is `DataType::Timestamp(TimeUnit::Nanosecond, _)`, because it is the only time unit supported by Arrow, which forces us to change the original time unit to nanoseconds. Although not ideal, this is not a problem, as both the left and right hand side of an expression are converted into `DataType::Timestamp(TimeUnit::Nanosecond, _)` and the expression can be properly evaluated.

# Are these changes tested?

Yes, corresponding test cases are added to the test function `test_type_coercion` from `datafusion/expr/src/type_coercion/binary.rs`.

# Are there any user-facing changes?

No.

### CODE REVIEW

File: datafusion/core/tests/sql/timestamp.rs [hunk 1]
```diff
@@ -1555,7 +1555,7 @@ async fn cast_timestamp_to_timestamptz() -> Result<()> {
 #[tokio::test]
 async fn test_cast_to_time() -> Result<()> {
     let ctx = SessionContext::new();
-    let sql = "SELECT 0::TIME";
+    let sql = "SELECT 0::TIME64";
```
Comment (waitingkuo) [inline datafusion/core/tests/sql/timestamp.rs:1558]: i'm not sure why do we have this. but `select 0::time64` doesn't work for me

File: datafusion/proto/proto/datafusion.proto [hunk 2]
```diff
@@ -830,10 +844,12 @@ message ScalarValue{
         ScalarDictionaryValue dictionary_value = 27;
         bytes binary_value = 28;
         bytes large_binary_value = 29;
+
         ScalarTime64Value time64_value = 30;
         IntervalMonthDayNanoValue interval_month_day_nano = 31;
         StructValue struct_value = 32;
         ScalarFixedSizeBinary fixed_size_binary_value = 34;
+>>>>>>> upstream/master
```
Comment (waitingkuo) [inline datafusion/proto/proto/datafusion.proto:852]:
```suggestion
```

File: datafusion/proto/proto/datafusion.proto [hunk 3]
```diff
@@ -770,6 +770,20 @@ message ScalarTimestampValue {
...
+  };
+}
+
+message ScalarTime64Value {
+  oneof value {
+    int64 time64_microsecond_value = 1;
+    int64 time64_nanosecond_value = 2;
+  };
+}
+
```
Comment (waitingkuo) [inline datafusion/proto/proto/datafusion.proto:786]: i think you should rebase master, the pr for time32/64 is merged

File: datafusion/core/tests/sql/select.rs [hunk 4]
```diff
@@ -814,6 +814,8 @@ async fn query_on_string_dictionary() -> Result<()> {
     ];
     assert_batches_eq!(expected, &actual);
 
+    // filtering with Time32 and Time64 types
+
```
Comment (waitingkuo) [inline datafusion/core/tests/sql/select.rs:818]: not quite sure what it is, i guess we need to delete ths

**Review by waitingkuo — COMMENTED** [review]
thank you @andre-cc-natzka 

here's my comments:
1. rebase master, your time32/64 pr is merged, thanks again
2. i'll suggest align the text with the prettyprint result (what you see in datafusion-cli)
i.e.
align the result for
```bash
❯ select now()::text;
+-----------------------------------+
| now()                             |
+-----------------------------------+
| 2022-11-21 21:42:52.196874 +00:00 |
+-----------------------------------+
```
to
```bash
❯ select now();
+----------------------------------+
| now()                            |
+----------------------------------+
| 2022-11-21T21:42:48.800795+00:00 |
+----------------------------------+
```
and
```bash
❯ select now()::timestamp::text;
+----------------------------+
| now()                      |
+----------------------------+
| 2022-11-21 21:44:48.386022 |
+----------------------------+
1 row in set. Query took 0.003 seconds.
```
to
```
❯ select now()::timestamp;
+----------------------------+
| now()                      |
+----------------------------+
| 2022-11-21T21:44:52.195395 |
+----------------------------+
```
3. delete the generated datafusion.rs in the proto folder, it's generated by the prost crate `https://github.com/tokio-rs/prost`

**Comment by alamb** [comment]
Marking this PR as draft as it still needs some work prior to passing CI

**Review by alamb — APPROVED** [review]
The code looks good to me. Thank you @andre-cc-natzka 

I would recommend (perhaps as a follow on) to add some SQL level tests but since there is coverage here I think it looks good to me

**Comment by waitingkuo** [comment]
hi @andre-cc-natzka  thank you, it looks good to me.

I suggest that add some SQL level tests 

e.g.

```bash
❯ select (timestamp '2000-01-01T00:00:00') = '2000-01-01T00:00:00';
+-----------------------------------------------------------+
| Utf8("2000-01-01T00:00:00") = Utf8("2000-01-01T00:00:00") |
+-----------------------------------------------------------+
| true                                                      |
+-----------------------------------------------------------+
1 row in set. Query took 0.002 seconds.
```

and this should raise error until we could cast TimestampTz to Utf8

```bash
❯ select (timestamptz '2000-01-01T00:00:00+00:00') = '2000-01-01T00:00:00+08:00';
NotImplemented("Unsupported CAST from Utf8 to Timestamp(Nanosecond, Some(\"+00:00\"))")
```

**Comment by alamb** [comment]
I plan to write some SQL level tests for this feature as they will help us in IOx.

**Comment by andre-cc-natzka** [comment]
Hello again @alamb , @waitingkuo. Thank you very much for your second revision and for merging the PR!

Thank you @alamb also for adding the SQL tests that test this new functionality, they look really nice! 

All the best,
André.

**Comment by waitingkuo** [comment]
thank you @andre-cc-natzka

**Pull request merged**